rfc/src/rfcs/0010-ghost-state.md

```diff
@@ -0,0 +1,207 @@
+- **Feature Name:** Ghost State (`ghost-state`)
+- **Feature Request Issue:** [#3184](https://github.com/model-checking/kani/issues/3184)
+- **RFC PR:**
+- **Status:** Unstable
+- **Version:** 0
+- **Proof-of-concept:** N/A
+
+-------------------
+
+## Summary
+
+Add support to ghost state to allow users to track metadata useful for verification purpose without extra
+runtime overhead.
+
+## User Impact
+
+TODO
+
+## User Experience
+
+We propose that ghost state to be implemented using a trait which is extended by Kani.
+
+```rust
+/// Trait that indicates a type has an associated ghost code of type `S`.
+///
+/// A type can have many multiple implementations, each will be mapped to a different ghost state.
+pub trait GhostState<S: Default + GhostValue>: Sized + GhostStateExt<S> {}
+
+/// Trait that specifies types that are safe to be used as a ghost value.
+///
+/// # Safety
+///
+/// A ghost value has to conform to the following existing restrictions:
+/// - Type size is less or equal to 8 bits (this may become a compilation error in the future).
+/// - The byte `0` is a valid representation of this type.
+/// - A ghost state is always initialized with the byte `0`.
+pub unsafe trait GhostValue: Sized {}
+
+/// These functions are automatically implemented  
+pub trait GhostStateExt<S> {
+    /// Set the value of a ghost state.
+    fn set_ghost_state(&self, value: S);
+
+    /// Get the value of the ghost state.
+    fn ghost_state(&self) -> S;
+
+    /// Add a dummy function with a private structure so this trait cannot be overwritten.
+    #[doc(hidden)]
+    fn private(&self, _: Internal);
+}
+```
+
+For example, let's say a user wants to check their union have been initialized, together with
+which variant.
+
+```rust
+union Size {
```

Comment:
The name is confusing (wrt to the trait "Sized")

Reply:
Sure. I can change it, this is just a dummy example.